File: runner_scripts/bench_key.jl
# The dataset key "<os>_<gpu>" of this machine, as bench_key.py computes it: the GPU
# name from nvidia-smi, tokenised on non-alphanumeric characters, the "NVIDIA"/"GeForce"
# vendor words dropped, the rest joined with '-'.
# e.g. "NVIDIA GeForce RTX 2060 SUPER" -> "RTX-2060-SUPER".

function _gpu_name_raw()
    try
        raw = read(`nvidia-smi --query-gpu=name --format=csv,noheader`, String)
        for line in split(raw, '\n')
            s = strip(line)
            isempty(s) || return String(s)
        end
    catch
    end
    return ""
end

function _os_key()
    Sys.iswindows() && return "windows"
    Sys.isapple()   && return "macos"
    Sys.islinux()   && return "linux"
    return "unknown"
end

function _sanitize_gpu(raw)
    tokens = filter(t -> !isempty(t) && t != "NVIDIA" && t != "GeForce",
                    split(raw, r"[^A-Za-z0-9]+"))
    isempty(tokens) ? "unknown-gpu" : join(tokens, "-")
end

"Return \"<os>_<gpu>\" for this machine."
dataset_key() = string(_os_key(), "_", _sanitize_gpu(_gpu_name_raw()))

File: runner_scripts/bench_key.py
"""The dataset key "<os>_<gpu>" of this machine: the nvidia-smi GPU name tokenised on non-alphanumerics, "NVIDIA"/"GeForce" dropped, joined with '-' (RTX-2060-SUPER); `python bench_key.py` prints it."""

import platform
import re
import subprocess


def _gpu_name_raw():
    try:
        out = subprocess.run(
            ["nvidia-smi", "--query-gpu=name", "--format=csv,noheader"],
            capture_output=True, text=True, timeout=15,
        )
        if out.returncode == 0:
            for line in out.stdout.splitlines():
                line = line.strip()
                if line:
                    return line
    except Exception:
        pass
    return ""


def _sanitize_gpu(raw):
    tokens = [t for t in re.split(r"[^A-Za-z0-9]+", raw)
              if t and t not in ("NVIDIA", "GeForce")]
    return "-".join(tokens) if tokens else "unknown-gpu"


def _os_key():
    return {"Linux": "linux", "Darwin": "macos", "Windows": "windows"}.get(
        platform.system(), "unknown")


def dataset_key():
    """Return "<os>_<gpu>" for this machine."""
    return "{0}_{1}".format(_os_key(), _sanitize_gpu(_gpu_name_raw()))


if __name__ == "__main__":
    print(dataset_key())
